### ISSUE TITLE
Derive `Hash` for types that are valuable as keys

### ISSUE THREAD

**Issue by czy-29** [gfx-rs/wgpu#6834]
Derive `Hash` for common types like: `DeviceType`, `AdapterInfo`, etc.

**Comment by jimblandy** [comment]
I don't think we want to just reflexively throw `Hash` and `PartialEq` on all "common types", but certainly for those two types it makes sense to use them as keys.
